lib/mssql.js

```diff
@@ -0,0 +1,215 @@
+  additionalProperties: false,
+  required: ["sql"],
+  properties: {
+    sql: {type: "string", minLength: 1},
+    params: {type: "array"},
+  },
+});
+
+// Since we want to work with more than one DB,
+// we will need to use a custom version of the MSSQL pool.
+// See: https://tediousjs.github.io/node-mssql/#connection-pools
+export const mssqlPool = {
+  get: (name, config) => {
+    if (!pools.has(name)) {
+      if (!config) {
+        throw new Error("Pool does not exist");
+      }
+
+      const pool = new mssql.ConnectionPool(config);
+      const close = pool.close.bind(pool);
+      pool.close = (...args) => {
+        pools.delete(name);
+        return close(...args);
+      };
+
+      pools.set(name, pool.connect());
+    }
+
+    return pools.get(name);
+  },
+
+  closeAll: () =>
+    Promise.all(
+      Array.from(pools.values()).map((connect) => {
+        return connect.then((pool) => pool.close());
+      })
+    ),
+};
+
+export function sanitizeForCellTag(string) {
+  let idx = 0;
+  const sanitized = string.split("").reduce((acc, char) => {
+    if (char === "?") {
+      idx += 1;
+      return acc.concat(`@${idx}`);
+    }
+    return acc.concat(char);
+  }, []);
+
+  return sanitized.join("");
+}
+
+export async function query(req, res, pool) {
+  const db = await pool;
+  const body = await json(req);
+
+  if (!validate(body)) throw badRequest();
+
+  res.setHeader("Content-Type", "text/plain");
+  const keepAlive = setInterval(() => res.write("\n"), 25e3);
+
+  let {sql, params = []} = body;
+  sql = sanitizeForCellTag(sql);
+
+  try {
+    await new Promise((resolve, reject) => {
+      const request = new mssql.Request(db);
+      const stream = request.toReadableStream();
+
+      params.forEach((param, idx) => {
+        request.input(`${idx + 1}`, param);
+      });
+
+      request.query(sql);
+      request.once("recordset", () => clearInterval(keepAlive));
+      request.on("recordset", (columns) => {
+        const schema = {
+          type: "array",
+          items: {
+            type: "object",
+            properties: Object.entries(columns).reduce(
+              (schema, [name, props]) => {
+                return {
+                  ...schema,
+                  ...{[name]: dataTypeSchema({type: props.type.name})},
+                };
+              },
+              {}
+            ),
+          },
+        };
+
+        res.write(`${JSON.stringify(schema)}`);
+        res.write("\n");
+      });
+
+      stream.pipe(JSONStream.stringify("", "\n", "\n")).pipe(res);
+      stream.on("done", () => {
+        resolve();
+      });
+      stream.on("error", (error) => {
+        if (!request.canceled) {
+          request.cancel();
+        }
+        reject(error);
+      });
+    });
+  } catch (error) {
+    if (!error.statusCode) error.statusCode = 400;
+    throw error;
+  } finally {
+    clearInterval(keepAlive);
+  }
+
+  res.end();
+}
+
+export async function check(req, res, pool) {
```

Comment:
I think that previously we didn't do this check at all for the database proxy. For people setting up the proxy, I guess we didn't care if they used a database user that had a lot of permissions. I'm wondering if we really want to require that you use a read-only user with Observable? Maybe something for a broader discussion.